gadopt/time_stepper.py

```diff
@@ -117,120 +124,209 @@ def advance(
+        mesh = solution.ufl_domain()
+        mc = MeshConstant(mesh)
+        self.t = mc.Constant(0.0)
+        self.dt_mesh_const = mc.Constant(float(dt))  # Internal MeshConstant for Irksome
+
+        # Build the Irksome form using the unified irksome_form method
+        # This ensures solution-dependent coefficients in the mass term are
+        # correctly evaluated at the current stage solution
+        F = equation.irksome_form(solution, Dt)
+
+        # Create the Irksome time stepper using MeshConstant
+        # For stage_type="deriv", pass bc_type parameter
+        irksome_kwargs = {
+            "stage_type": stage_type,
+            "bcs": strong_bcs or (),
+            "solver_parameters": solver_parameters or {}
+        }
+
+        # Add bc_type only for stage formulations that support it
+        if stage_type == "deriv":
+            irksome_kwargs["bc_type"] = bc_type
+
+        self.stepper = IrksomeTimeStepper(
+            F,
+            butcher,
+            self.t,
+            self.dt_mesh_const,  # Use MeshConstant for Irksome
+            solution,
+            **irksome_kwargs
         )
+
+        self.update_forcings = None
         self._initialized = False
-        V = solution.function_space()
-        assert V == equation.trial_space
 
-        self.tendency = []
-        for i in range(self.n_stages):
-            k = firedrake.Function(V, name='tendency{:}'.format(i))
-            self.tendency.append(k)
-
-        # fully explicit evaluation
-        trial = firedrake.TrialFunction(V)
-        self.a_rk = self.equation.mass(trial)
-        self.l_rk = self.dt_const * self.equation.residual(self.solution)
-
-        self._nontrivial = self.l_rk != 0
-
-        # construct expressions for stage solutions
-        if self._nontrivial:
-            self.sol_expressions = []
-            for i_stage in range(self.n_stages):
-                sol_expr = sum(map(operator.mul, self.tendency[:i_stage], self.a[i_stage][:i_stage]))
-                self.sol_expressions.append(sol_expr)
-            self.final_sol_expr = sum(map(operator.mul, self.tendency, self.b))
-
-        self.update_solver()
-
-    def update_solver(self):
-        """Create solver objects"""
-        if self._nontrivial:
-            self.solver = []
-            for i in range(self.n_stages):
-                prob = firedrake.LinearVariationalProblem(self.a_rk, self.l_rk, self.tendency[i], bcs=self.hom_bcs)
-                solver = firedrake.LinearVariationalSolver(prob, options_prefix=self.name + '_k{:}'.format(i),
-                                                           solver_parameters=self.solver_parameters)
-                self.solver.append(solver)
-
-    def initialize(self, solution):
-        self.solution_old.assign(solution)
+    def initialize(self, init_solution):
+        """Initialize the time integrator."""
+        self.solution.assign(init_solution)
         self._initialized = True
```

Comment:
We should remove any initialisation happening outside class instantiation, like we did for `stokes_integrators.py`.